File: backend/app/services/resume_text_builder.py
from typing import Dict, Any, List


class ResumeTextBuilder:
    """
    Unified Resume Text Builder.
    Converts structured resume data (from MongoDB, uploaded parser, or parsed JSON)
    into a standardized, ATS-friendly plain text format for AI analysis and matching.
    """

    @staticmethod
    def build_ats_text(resume: Dict[str, Any]) -> str:
        if not isinstance(resume, dict):
            return ""

        sections: List[str] = []

        # 1. Personal Information & Header
        personal = resume.get("personal_info") or {}
        if not isinstance(personal, dict):
            personal = {}

        name = personal.get("full_name") or resume.get("title") or "Candidate"
        target_role = resume.get("target_role") or ""
        email = personal.get("email") or ""
        phone = personal.get("phone") or ""
        location = personal.get("location") or ""
        website = personal.get("website") or ""
        github = personal.get("github") or ""
        linkedin = personal.get("linkedin") or ""

        header_lines = [name.upper()]
        if target_role:
            header_lines.append(target_role.upper())

        contact_items = [item for item in [email, phone, location, website, github, linkedin] if item]
        if contact_items:
            header_lines.append(" | ".join(contact_items))

        sections.append("\n".join(header_lines))

        # 2. Professional Summary
        summary = personal.get("summary") or resume.get("summary") or ""
        if summary and str(summary).strip():
            sections.append(f"PROFESSIONAL SUMMARY\n{str(summary).strip()}")

        # 3. Work Experience
        work_items = resume.get("work_experience") or resume.get("experience") or []
        if isinstance(work_items, list) and work_items:
            exp_lines = ["WORK EXPERIENCE"]
            for exp in work_items:
                if not isinstance(exp, dict):
                    continue
                pos = exp.get("position") or exp.get("title") or "Role"
                comp = exp.get("company") or exp.get("organization") or ""
                duration = exp.get("duration") or exp.get("dates") or ""
                exp_loc = exp.get("location") or ""

                title_line = f"{pos} - {comp}".strip(" - ")
                if duration:
                    title_line = f"{title_line} ({duration})"
                exp_lines.append(title_line)

                if exp_loc:
                    exp_lines.append(exp_loc)

                bullets = exp.get("bullets") or exp.get("description") or []
                if isinstance(bullets, str):
                    bullets = [bullets]
                elif isinstance(bullets, list):
                    bullets = [str(b) for b in bullets if str(b).strip()]

                for b in bullets:
                    exp_lines.append(f"- {b}")
                exp_lines.append("")  # empty line separator

            sections.append("\n".join(exp_lines).strip())

        # 4. Skills
        skills = resume.get("skills") or []
        if isinstance(skills, list) and skills:
            skill_lines = ["SKILLS"]
            for cat in skills:
                if isinstance(cat, dict):
                    cat_name = cat.get("category") or "Technical Skills"
                    items = cat.get("items") or []
                    if isinstance(items, list):
                        items_str = ", ".join([str(i) for i in items if str(i).strip()])
                        if items_str:
                            skill_lines.append(f"{cat_name}: {items_str}")
                elif isinstance(cat, str) and cat.strip():
                    skill_lines.append(f"- {cat.strip()}")

            sections.append("\n".join(skill_lines).strip())

        # 5. Projects
        projects = resume.get("projects") or []
        if isinstance(projects, list) and projects:
            proj_lines = ["PROJECTS"]
            for proj in projects:
                if not isinstance(proj, dict):
                    continue
                proj_name = proj.get("name") or proj.get("title") or "Project"
                proj_link = proj.get("link") or proj.get("url") or ""
                proj_desc = proj.get("description") or proj.get("summary") or ""
                tech = proj.get("tech_stack") or proj.get("technologies") or []

                p_header = proj_name
                if proj_link:
                    p_header = f"{proj_name} ({proj_link})"
                proj_lines.append(p_header)

                if proj_desc:
                    proj_lines.append(f"- {proj_desc}")

                if isinstance(tech, list) and tech:
                    tech_str = ", ".join([str(t) for t in tech if str(t).strip()])
                    if tech_str:
                        proj_lines.append(f"  Technologies: {tech_str}")
                proj_lines.append("")

            sections.append("\n".join(proj_lines).strip())

        # 6. Education
        education = resume.get("education") or []
        if isinstance(education, list) and education:
            edu_lines = ["EDUCATION"]
            for edu in education:
                if not isinstance(edu, dict):
                    continue
                degree = edu.get("degree") or "Degree"
                field = edu.get("field_of_study") or edu.get("major") or ""
                inst = edu.get("institution") or edu.get("university") or edu.get("school") or ""
                edu_dur = edu.get("duration") or edu.get("dates") or ""
                grade = edu.get("grade") or edu.get("gpa") or ""

                deg_str = degree
                if field and field.lower() not in degree.lower():
                    deg_str = f"{degree} in {field}"

                edu_header = f"{deg_str} - {inst}".strip(" - ")
                if edu_dur:
                    edu_header = f"{edu_header} ({edu_dur})"
                edu_lines.append(edu_header)

                if grade:
                    edu_lines.append(f"Grade / GPA: {grade}")
                edu_lines.append("")

            sections.append("\n".join(edu_lines).strip())

        # 7. Certifications
        certificates = resume.get("certificates") or resume.get("certifications") or []
        if isinstance(certificates, list) and certificates:
            cert_lines = ["CERTIFICATIONS"]
            for cert in certificates:
                if not isinstance(cert, dict):
                    continue
                cert_name = cert.get("name") or cert.get("title") or "Certification"
                issuer = cert.get("issuer") or cert.get("organization") or ""
                date = cert.get("date") or cert.get("year") or ""

                c_line = cert_name
                if issuer:
                    c_line = f"{c_line} - {issuer}"
                if date:
                    c_line = f"{c_line} ({date})"
                cert_lines.append(f"- {c_line}")

            sections.append("\n".join(cert_lines).strip())

        return "\n\n".join([s for s in sections if s.strip()])

File: frontend/src/utils/resumeTextBuilder.ts
/**
 * Client-Side Resume ATS Text Builder.
 * Converts structured resume data into standardized ATS plain text.
 */
export function buildResumeAtsText(resume: any): string {
  if (!resume || typeof resume !== 'object') return ''

  const sections: string[] = []

  // 1. Personal Information & Header
  const personal = resume.personal_info || {}
  const name = personal.full_name || resume.title || 'Candidate'
  const targetRole = resume.target_role || ''
  const email = personal.email || ''
  const phone = personal.phone || ''
  const location = personal.location || ''
  const website = personal.website || ''
  const github = personal.github || ''
  const linkedin = personal.linkedin || ''

  const headerLines: string[] = [name.toUpperCase()]
  if (targetRole) headerLines.push(targetRole.toUpperCase())

  const contactItems = [email, phone, location, website, github, linkedin].filter(Boolean)
  if (contactItems.length > 0) {
    headerLines.push(contactItems.join(' | '))
  }

  sections.push(headerLines.join('\n'))

  // 2. Professional Summary
  const summary = personal.summary || resume.summary || ''
  if (summary && summary.trim()) {
    sections.push(`PROFESSIONAL SUMMARY\n${summary.trim()}`)
  }

  // 3. Work Experience
  const workItems = resume.work_experience || resume.experience || []
  if (Array.isArray(workItems) && workItems.length > 0) {
    const expLines: string[] = ['WORK EXPERIENCE']
    for (const exp of workItems) {
      if (!exp || typeof exp !== 'object') continue
      const pos = exp.position || exp.title || 'Role'
      const comp = exp.company || exp.organization || ''
      const duration = exp.duration || exp.dates || ''
      const expLoc = exp.location || ''

      let titleLine = [pos, comp].filter(Boolean).join(' - ')
      if (duration) titleLine += ` (${duration})`
      expLines.push(titleLine)

      if (expLoc) expLines.push(expLoc)

      const bullets = Array.isArray(exp.bullets)
        ? exp.bullets
        : typeof exp.bullets === 'string'
        ? [exp.bullets]
        : []

      for (const b of bullets) {
        if (b && String(b).trim()) {
          expLines.push(`- ${String(b).trim()}`)
        }
      }
      expLines.push('')
    }
    sections.push(expLines.join('\n').trim())
  }

  // 4. Skills
  const skills = resume.skills || []
  if (Array.isArray(skills) && skills.length > 0) {
    const skillLines: string[] = ['SKILLS']
    for (const cat of skills) {
      if (cat && typeof cat === 'object') {
        const catName = cat.category || 'Technical Skills'
        const items = Array.isArray(cat.items) ? cat.items.filter(Boolean) : []
        if (items.length > 0) {
          skillLines.push(`${catName}: ${items.join(', ')}`)
        }
      } else if (typeof cat === 'string' && cat.trim()) {
        skillLines.push(`- ${cat.trim()}`)
      }
    }
    sections.push(skillLines.join('\n').trim())
  }

  // 5. Projects
  const projects = resume.projects || []
  if (Array.isArray(projects) && projects.length > 0) {
    const projLines: string[] = ['PROJECTS']
    for (const proj of projects) {
      if (!proj || typeof proj !== 'object') continue
      const projName = proj.name || proj.title || 'Project'
      const projLink = proj.link || proj.url || ''
      const projDesc = proj.description || proj.summary || ''
      const tech = Array.isArray(proj.tech_stack)
        ? proj.tech_stack.filter(Boolean)
        : []

      let pHeader = projName
      if (projLink) pHeader += ` (${projLink})`
      projLines.push(pHeader)

      if (projDesc) projLines.push(`- ${projDesc}`)
      if (tech.length > 0) projLines.push(`  Technologies: ${tech.join(', ')}`)
      projLines.push('')
    }
    sections.push(projLines.join('\n').trim())
  }

  // 6. Education
  const education = resume.education || []
  if (Array.isArray(education) && education.length > 0) {
    const eduLines: string[] = ['EDUCATION']
    for (const edu of education) {
      if (!edu || typeof edu !== 'object') continue
      const degree = edu.degree || 'Degree'
      const field = edu.field_of_study || edu.major || ''
      const inst = edu.institution || edu.university || edu.school || ''
      const eduDur = edu.duration || edu.dates || ''
      const grade = edu.grade || edu.gpa || ''

      let degStr = degree
      if (field && !degree.toLowerCase().includes(field.toLowerCase())) {
        degStr = `${degree} in ${field}`
      }

      let eduHeader = [degStr, inst].filter(Boolean).join(' - ')
      if (eduDur) eduHeader += ` (${eduDur})`
      eduLines.push(eduHeader)

      if (grade) eduLines.push(`Grade / GPA: ${grade}`)
      eduLines.push('')
    }
    sections.push(eduLines.join('\n').trim())
  }

  // 7. Certifications
  const certs = resume.certificates || resume.certifications || []
  if (Array.isArray(certs) && certs.length > 0) {
    const certLines: string[] = ['CERTIFICATIONS']
    for (const cert of certs) {
      if (!cert || typeof cert !== 'object') continue
      const certName = cert.name || cert.title || 'Certification'
      const issuer = cert.issuer || cert.organization || ''
      const date = cert.date || cert.year || ''

      let cLine = certName
      if (issuer) cLine += ` - ${issuer}`
      if (date) cLine += ` (${date})`
      certLines.push(`- ${cLine}`)
    }
    sections.push(certLines.join('\n').trim())
  }

  return sections.filter((s) => s.trim().length > 0).join('\n\n')
}
